docs/architecture/adr-022-custom-panic-handling.md

```diff
@@ -0,0 +1,208 @@
+# ADR 022: Custom BaseApp panic handling
+
+## Changelog
+
+- 2020 Apr 24: Initial Draft
+
+## Context
+
+The current implementation of BaseApp does not allow developers to write custom error handlers during panic recovery
+[runTx()](https://github.com/cosmos/cosmos-sdk/blob/bad4ca75f58b182f600396ca350ad844c18fc80b/baseapp/baseapp.go#L539)
+method. We think that this method can be more flexible and can give SDK users more options for customizations without
+the need to rewrite whole BaseApp. Also there's one special case for `sdk.ErrorOutOfGas` error handling, that case
+might be handled in a "standard" way (middleware) alongside the others.
+
+We propose middleware-solution, which could help developers implement the following cases:
+* add external logging (let's say sending reports to external services like [Sentry](https://sentry.io));
+* call panic for specific error cases;
```

Comment:
Like what? Why would you want to panic during transaction execution instead of reverting? Can you give an example?

Reply:
https://github.com/cosmos/cosmos-sdk/pull/6072#discussion_r414756015

Reply:
Can you add that as a written (more detailed) example?

Reply:
I don't get you. You want me to add our use case as an example to the ADR?

Reply:
Yes, please do.

Reply:
Added: https://github.com/cosmos/cosmos-sdk/pull/6072/commits/bf1cbdf259bebd624a74bb28360de3b131c74e53